### PR TITLE
Fix previous link relationship question

### PR DESCRIPTION
**Pull request by LJBabbage**
### What is the context of this PR?
The household relationship grid questions in census needs to have a previous button, both on the first block (returns to overnight-visitors) and any repeating blocks (instance before). Although not ideal after talking to Andrew we decided to hard code it for now

### How to review 
With Census, fill in multiple people (>2) on the who lives here page, navigate to relationship grid, making sure the first block and all repeating blocks have a previous working button

### CODE REVIEW

File: app/views/questionnaire.py [hunk 1]
```diff
@@ -491,10 +491,20 @@ def _build_template(current_location, context=None, template=None):
 
     if previous_location is not None and not is_first_block_for_group and not current_location.block_id == 'thank-you':
         previous_url = previous_location.url(metadata)
+    elif current_location.group_id == 'who-lives-here-relationship':
+        previous_url = _relationship_grid_previous_url(current_location.group_instance).url(metadata)
 
     return _render_template(context, current_location.block_id, front_end_navigation, metadata_context, current_location, previous_url, template)
 
 
+def _relationship_grid_previous_url(current_instance):
```
Comment (iwootten) [inline app/views/questionnaire.py]: I don't think we need to have "_grid" in the method name here, we don't use it elsewhere and there's only 1 type of relationship page. 

It might also make sense to clarify the type of instance that is passed in eg. current_group_instance, given we have other types of instances in the system.

File: tests/integration/questionnaire/test_questionnaire_previous_link.py [hunk 2]
```diff
@@ -74,3 +74,43 @@ def test_previous_link_doesnt_appear_on_thank_you(self):
         self.assertTrue(final_url.endswith('thank-you'))
         self.assertNotIn('Previous', content)
 
+    def test_previous_link_on_relationship_grid(self):
```
Comment (iwootten) [inline tests/integration/questionnaire/test_questionnaire_previous_link.py]: Similarly on the "_grid" method naming here.

**Review by iwootten — CHANGES_REQUESTED** [review]

File: app/views/questionnaire.py (hunk 1 above)
Comment (benjefferies) [reply app/views/questionnaire.py]: It might be nice to have one method to build the previous url. I think it would make the behemoth _build_template method nicer too

**Review by iwootten — APPROVED** [review]

File: app/views/questionnaire.py [hunk 3]
```diff
@@ -491,10 +491,20 @@ def _build_template(current_location, context=None, template=None):
 
     if previous_location is not None and not is_first_block_for_group and not current_location.block_id == 'thank-you':
         previous_url = previous_location.url(metadata)
+    elif current_location.group_id == 'who-lives-here-relationship':
```
Comment (benjefferies) [inline app/views/questionnaire.py]: Is this the right place to fix this issue? Shouldn't the previous location generated by the path_finder get the previous group instance?

**Review by benjefferies — CHANGES_REQUESTED** [review]

File: app/views/questionnaire.py (hunk 3 above)
Comment (LJBabbage) [reply app/views/questionnaire.py]: Will need to take is_first_block_for_group or current_location.block_id == 'thank-you' with me to do that, although I don't see a problem with that, potentially they should be down there anyway

**Comment by benjefferies** [comment]
Not sure if the change fully reflects what I mean. Shouldn't the current - 1 of get_location_path path be the previous group instance?

**Comment by LJBabbage** [comment]
We just can't do that, that was our original problem and why we never had previous on the first block of a group. If they jump around the survey using navigation the previous group (-1) might not be the group they were on before.

**Comment by LJBabbage** [comment]
Updated. One test was removed as I realised it was already covered. The other one was misleading. The path_finder.get_previous_location was returning a value which was then being hidden by the view, so it didn't reflect what was actually happening. We don't put a link to the introduction from any block on the front-end so why assert we do.

File: app/questionnaire/path_finder.py [hunk 4]
```diff
@@ -243,3 +252,11 @@ def get_latest_location(self, completed_blocks=None):
                 return incomplete_blocks[0]
 
         return location_path[0]
+
+    @staticmethod
+    def _relationship_previous_url(current_group_instance):
```
Comment (benjefferies) [inline app/questionnaire/path_finder.py]: Should be updated to reflect it's not a url

File: app/questionnaire/path_finder.py [hunk 5]
```diff
@@ -243,3 +252,11 @@ def get_latest_location(self, completed_blocks=None):
...
+        if current_group_instance == 0:
+            previous_url = Location('who-lives-here', 0, 'overnight-visitors')
```
Comment (benjefferies) [inline app/questionnaire/path_finder.py]: Should be previous_location

File: tests/app/questionnaire/test_path_finder.py [hunk 6]
```diff
@@ -217,7 +217,7 @@ def test_get_previous_location_introduction(self):
 
         previous_location = path_finder.get_previous_location(current_location=first_location)
 
-        self.assertEqual('introduction', previous_location.block_id)
+        self.assertEqual(None, previous_location)
```
Comment (benjefferies) [inline tests/app/questionnaire/test_path_finder.py]: Why has this test changed? Looks unrelated

File: tests/app/questionnaire/test_path_finder.py [hunk 7]
```diff
@@ -519,18 +519,6 @@ def test_repeating_groups_no_of_answers_minus_one(self):
 
         self.assertEqual(expected_path, path_finder.get_routing_path())
 
-    def test_repeating_groups_previous_location_introduction(self):
```
Comment (benjefferies) [inline tests/app/questionnaire/test_path_finder.py:522]: I think this test is still valid?

**Review by benjefferies — CHANGES_REQUESTED** [review]

File: tests/app/questionnaire/test_path_finder.py (hunk 7 above)
Comment (LJBabbage) [inline tests/app/questionnaire/test_path_finder.py]: See note

File: tests/app/questionnaire/test_path_finder.py (hunk 6 above)
Comment (LJBabbage) [reply tests/app/questionnaire/test_path_finder.py]: See note

**Review by benjefferies — APPROVED** [review]
